### PR TITLE
transpose(MatMul(<Symbol>, Matrix)) now works properly

### PR DESCRIPTION
**Pull request by gxyd**
I am not sure if this is the fix. But it seems to be working for me, i guess.

fixes issue #9503 
Please have a look

### CODE REVIEW

File: sympy/matrices/expressions/matmul.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from sympy.core import Mul, Basic, sympify, Add
 from sympy.core.compatibility import range
 from sympy.functions import adjoint
-from sympy.matrices.expressions.transpose import transpose
+from sympy.functions.elementary.complexes import transpose
```
Comment (toolforger) [inline sympy/matrices/expressions/matmul.py]: That's what I call "experimental programming": If the function that's called doesn't work, let's hope another, similarly-named one from another module will.
Except it's just fixing symptoms unless the other function is truly the one that should have been called - which is extremely unlikely because we're inside `matrices`.
Also, I doubt it's even fixing symptoms. Try adding a test that transposes a 4x3 matrix; I'm pretty sure that this will fail.

File: sympy/matrices/expressions/tests/test_matmul.py [hunk 2]
```diff
@@ -119,3 +119,9 @@ def test_collapse_MatrixBase():
 
 def test_refine():
     assert refine(C*C.T*D, Q.orthogonal(C)).doit() == D
+
+def test_issue_9503():
+    from sympy import I
+    a = symbols('a')
+    M = Matrix([[1, 2 + I], [3, 4]])
+    assert transpose(MatMul(a, M)) == MatMul(transpose(a), transpose(M))
```
Comment (toolforger) [inline sympy/matrices/expressions/tests/test_matmul.py]: This should be `MatMul(transpose(M), transpose(a))`.
I don't know why the above works, maybe the values just happen to coincide. Or maybe the definition of `transpose` for complex numbers is different than the one for matrices.

Comment (toolforger) [inline sympy/matrices/expressions/tests/test_matmul.py]: A test to add:

```
M1 = Matrix([[11, 12, 13], [21, 22, 23], [31, 32, 33], [41, 42, 43]])
M2 = Matrix([[11, 21, 31, 41], [12, 22, 32, 42], [13, 23, 33, 43]])
assert transpose(MatMul(M1, M2)) == MatMul(transpose(M2), transpose(M1))

**Comment by toolforger** [comment]
Ok... a Wikipedia lookup told me that the `complex transpose` would more properly be named `transjugate`, because it's a transpose combined with conjugating all elements.
Transjugating reverses factors, i.e. `tr(A*B) == tr(B) * tr(A)` where `tr` is the transjugating `transpose`.

**Comment by gxyd** [comment]
Please have a look.
I ran the following. Though i will add the tests to the my commits but i get the following using my current branch. Though no errors so far, are these you though may not pass.
Let me know if you think others may fails. Or i did not included ones you think may fail.
```
>>> M = Matrix([ [1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]  ])
>>> MT = transpose(M)
>>> V = Matrix([[1, 5, 9], [2, 6, 10], [3, 7, 11], [4, 8, 12]])
>>> MT == V
True
>>> a = Symbol( 'a' )
>>> Q = transpose( MatMul( a, M ) )
>>> P = MatMul( transpose(a), MT )
>>> Q == P
True
>>> M = Matrix([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12] ])
>>> a = Symbol( 'a' )
>>> assert transpose(MatMul(a, M)) == MatMul(transpose(a), transpose(M))   # no error raised
```

**Comment by gxyd** [comment]
> A test to add:
M1 = Matrix([[11, 12, 13], [21, 22, 23], [31, 32, 33], [41, 42, 43]])
M2 = Matrix([[11, 21, 31, 41], [12, 22, 32, 42], [13, 23, 33, 43]])
assert transpose(MatMul(M1, M2)) == MatMul(transpose(M2), transpose(M1))

Above `raises error`
But Following works
```
>>> assert transpose(MatMul(M1, M2)).doit() == MatMul(transpose(M2), transpose(M1)).doit()
```

**Comment by toolforger** [comment]
Transpose should invert the multiplication order.
I'm not sure what actually happens here, but something is really fishy here; we should at least understand what's happening before we consider anything fixed.

**Comment by toolforger** [comment]
Oh right I see what's happening in the interactive session - `a` is a scalar symbol. Scalar-to-matrix multiplication is indeed commutative, so the test cannot check that transposition properly inverts multiplication order.
I forgot all about `doit`, I guess it needs to be used to turn an unevaluated matrix multiplication into a matrix. I don't know why the two are different.

**Comment by gxyd** [comment]
It is just `Integral` and `Integral.doit()` i think.

**Comment by gxyd** [comment]
Then why MatMul(a, M) != MatMul(M, a) as for you said Matrix mul is commutative for scalars ? Look.

**Comment by toolforger** [comment]
I'm not sure what's actually happening here; I can't easily check myself right now because I'm in the process of reinstalling my flaky Python 2.x environments.
If I could check, I'd look not just at equalities but also at the expressions themselves, trying to `doit` or `eval` them to see what actually comes out of the expressions being compared, or maybe debug the equality processing itself to see what SymPy is actually doing.

**Comment by cbm755** [comment]
I think fix will be in `Transpose.doit()` method, or more likely in `_eval_transpose()` method of `MatMul`.  See recently fixes to trace, etc.

**Comment by toolforger** [comment]
@gxyd I got Python running again, so I can now answer:

```
>>> MatMul(M1, x)
⎡11  12  13⎤⋅x
⎢          ⎥  
⎢21  22  23⎥  
⎢          ⎥  
⎢31  32  33⎥  
⎢          ⎥  
⎣41  42  43⎦  
>>> MatMul(x, M1)
x⋅⎡11  12  13⎤
  ⎢          ⎥
  ⎢21  22  23⎥
  ⎢          ⎥
  ⎢31  32  33⎥
  ⎢          ⎥
  ⎣41  42  43⎦
```

However:

```
>>> M1 * x
⎡11⋅x  12⋅x  13⋅x⎤
⎢                ⎥
⎢21⋅x  22⋅x  23⋅x⎥
⎢                ⎥
⎢31⋅x  32⋅x  33⋅x⎥
⎢                ⎥
⎣41⋅x  42⋅x  43⋅x⎦
>>> x * M1
⎡11⋅x  12⋅x  13⋅x⎤
⎢                ⎥
⎢21⋅x  22⋅x  23⋅x⎥
⎢                ⎥
⎢31⋅x  32⋅x  33⋅x⎥
⎢                ⎥
⎣41⋅x  42⋅x  43⋅x⎦
```

and indeed I have:
```
>>> M1 * x == x * M1
True
```

I haven't verified, but I guess that `MatMul` simply constructs a multiplicative expression, but does not make any attempt at evaluation; I don't know which of `doit`, `eval`, or `simplify` is the right operation to call but whatever it is, it's being called by `*` and gives the expected answer.

(Lesson to take home: if two expressions do not compare equal that should be, double-check your assumptions about what is being compared.)

File: sympy/matrices/expressions/tests/test_matmul.py (hunk 2 above)
Comment (jksuom) [inline sympy/matrices/expressions/tests/test_matmul.py]: Perhaps one would prefer
```
transpose(MatMul(a, M)) == MatMul(a, transpose(M))
or
transpose(MatMul(a, M)) == MatMul(transpose(M), a)
```

Comment (toolforger) [inline sympy/matrices/expressions/tests/test_matmul.py]: I'm not seeing how that would be preferrable - can you elaborate?

Comment (jksuom) [inline sympy/matrices/expressions/tests/test_matmul.py]: Plain ``a`` instead of ``transpose(a)`` would seem more suitable for a scalar ``a``.

Comment (gxyd) [inline sympy/matrices/expressions/tests/test_matmul.py]: > Plain a instead of transpose(a) would seem more suitable for a scalar a

```
>>> from sympy.matrices.expressions.transpose import transpose as mat_trans
>>> from sympy.functions.elementary.complexes import transpose as func_trans
>>> a = Symbol('a')
>>> mat_trans(a) == a
False
>>> func_trans(a) == a
False
```
Why do you expect any of these print plain `a` instead of `mat_trans(a)` or `func_trans(a)` ?

Comment (jksuom) [inline sympy/matrices/expressions/tests/test_matmul.py]: If ``a`` is a scalar, it think that ``transpose(a)`` should mean the same as ``a`` itself. And I usually prefer the shortest expression, at least if no confusion would result.

Comment (gxyd) [inline sympy/matrices/expressions/tests/test_matmul.py]: first please properly specify which `transpose` you are talking about ? is it `matrices.expressions.transpose` for `functions.elementary.complexes.`

Comment (gxyd) [inline sympy/matrices/expressions/tests/test_matmul.py]:
```
>>> from sympy.matrices.expressions.transpose import transpose as mat_trans
>>> from sympy.functions.elementary.complexes import transpose as func_trans
>>> mat_trans(a).is_Matrix
True
>>> func_trans(a).is_Matrix
False
```
Now you see the two resulting in a different things.

Comment (gxyd) [inline sympy/matrices/expressions/tests/test_matmul.py]: As far you said `transpose(a)` should mean the same thing. In any case it should not
because for `a` we can later on substitute for values. And here `a` should not be referred to as being a scalar, i guess.
like.
```
>>> from sympy.matrices.expressions.transpose import transpose as mat_trans
>>> mat_trans(a).subs( { a: Matrix([ [1, 2, 3] ]) } )
```

Comment (jksuom) [inline sympy/matrices/expressions/tests/test_matmul.py]: > first please properly specify which transpose you are talking about ? 

I was talking about the transpose of a scalar (in the complex module). There does not even seem to exist a useful way of defining the transpose of a scalar in the matrix module.

Comment (jksuom) [inline sympy/matrices/expressions/tests/test_matmul.py]: > because for a we can later on substitute for values.

This is true, because ``subs`` does purely formal substitution. It does check the types of the substituted expressions. This can lead to problems. Instead of a scalar symbol ``a`` one should use ``MatrixSymbol``, if a matrix is to be substituted.

**Comment by gxyd** [comment]
Please now have a look. Now i have made a few changes. `transpose(a)` returns `a`.

**Comment by toolforger** [comment]
It's still failing.
You can try `bin/test name/of/test/file/to/run` before pushing to GitHub, that way you don't have to wait for Travis to see whether a specific change will work or not.
Given the current situation, it's possible that some test failures just document misunderstandings about `transpose`, but it's too many errors to assume that that's always the case.

**Comment by gxyd** [comment]
Though i knew that. But i was not expecting tests to fail in `physics` modules. Now all tests passed have on my machine.

File: sympy/core/tests/test_expr.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 from sympy.physics.secondquant import FockState
 from sympy.physics.units import meter
 from sympy.core.compatibility import range
+from sympy.functions import transpose
```
Comment (gxyd) [inline sympy/core/tests/test_expr.py]: Though there is no need for this i will remove this.

**Comment by jksuom** [comment]
I am beginning to suspect that the omission of ``Transpose`` in ``Transpose.doit`` would not be the proper fix.

If an object has an ``_eval_something`` attribute, it returns a suitable evaluation if it can find one. Otherwise it returns ``None`` in which case it is usually assumed that the caller would replace it with an unevaluated ``Something``. That is what happens in ``Transpose.doit`` if ``arg._eval_transpose`` returns ``None``. This principle is broken, if the bare ``arg`` is returned instead of ``Transpose(arg)``.

I think that the actual issue is that ``Expr._eval_transpose`` cannot evaluate the transpose of an ordinary symbol.
```
>>> a = Symbol('a')
>>> transpose(a)
transpose(a)
```
However, it knows what to do if the symbol is declared complex (which is usually true in connection with matrices).
```
>>> a = Symbol('a', complex=True)
>>> transpose(a)
a
```
Thus one solution would be to always use such symbols in computations with matrices.
But that may be considered too cumbersome.

That is why I (tentatively) suggest that the test ``self.is_complex`` in ``Expr._eval_transpose`` be replaced with ``self.is_commutative`` which would be automatically satisfied by ordinary symbols.

**Comment by gxyd** [comment]
Though i also noticed the this question on [SO] (http://stackoverflow.com/questions/22820000/how-to-solve-matrix-equation-with-sympy) , involving nature of `non-commutative symbols`. But i guess, what you suggest is not complete good, as for ex. 

ATM the problem is just that: "Why are we even allowing the`matrix transpose` to return unevaluated as `Transpose(a)` for non-commutative symbols ?"   As for you answered to this  as

> assumed that the caller would replace it with an unevaluated `Something`

I  think "according to you" that `Something` may be `Matrix` since you are keeping the `Transpose(a).doit()` unevaluated for `non-commutative symbols`. 

But why do we do that, even if we want to persform the `substitution` then why not just use the `MatrixSymbols` ?

```
>>> b = Symbol('b', commutative=False)
>>> Transpose(b).doit()
Transpose(b)        # this is what you suggest
>>> expr = b**2 + b
>>> expr.subs( { b: Matrix([ [1, 2], [3,4 ] ]) })      # is this what you called `Something` ?
```
I suggest
>>> A = MatrixSymbol( 'A', m, n)     # m  and n are integers
>>> Transpose( A ).doit()
Transpose(A)        # this is already OK in current code base
>>> expr = A**2 + A
>>> expr.subs( { A: Matrix([ [1, 2], [3,4 ] ]) })      # this is how to do substitution of Matrix better.
```

**Comment by jksuom** [comment]
```
>>> b = Symbol('b', commutative=False)
>>> Transpose(b).doit()
Transpose(b)        # this is what you suggest
```
I don't recall proposing to make ordinary symbols non-commutative. There are ``MatrixSymbols`` for that purpose.

**Comment by gxyd** [comment]
> That is why I (tentatively) suggest that the test self.is_complex in Expr._eval_transpose be replaced with self.is_commutative.

I tried this and found that `matrix_transpose` of `non-commutative Symbol` then does not have `MatMul` with the a `Matrix`. After changing 
```
>>> b = Symbol('b', commutative=False)
>>> MatMul( matrix_transpose(b), M )       # this is not defined then
AttributeError: 'Symbol' object has no attribute 'shape'
```
> I don't recall proposing to make ordinary symbols non-commutative.

Neither did i said so. Sorry if i could not make it clear. I only said that: "that property will be satisfied for non-commutative symbols  not talking about make them so.".

**Comment by jksuom** [comment]
> matrix_transpose of non-commutative Symbol then does not have MatMul with the a Matrix

That is true. It also seems that ``MatMul`` will accept a non-commutative symbol. However, it is not clear what that should mean as it has no shape. (In fact, I suspect that ``MatMul`` in its current form might have issues, because it fails to check the commutativity.)

Perhaps ``MatMul`` could be made to accept also transposes of non-commutative symbols, if that should really be needed.
 
(An alternative might be an implementation of the use of ``MatrixSymbol``s in ``solve``. That is what I would prefer, since the shape of the objects is then known all the time. I don't see any great obstacles to such an implementation, in principle.)

**Comment by gxyd** [comment]
I support changing to only Transpose.doit() the portion of `arg` instead of `Transpose(arg)` because
I think what i have done currently in my commits satisfies all the conditions properly. It should `not` depend upon whether a `symbol` is commutative or not. On what it should depend is whether we are taking a `matrix_transpose` or `complex_transform` and internal `functions applied` like is it after `MatMul` or `before MatMul`
```
>>> b = Symbol('b')     # it should not depend on whether it is commutative or not
>>> M = Matrix([ [1, 2], [3, 4] ])
>>> matrix_transpose( MatMul(b, M ) )
b * Matrix([ [1, 3], [2, 4] ])
>>>  complex_transpose( MatMul(b, M) )
b * Matrix([ [1, 3], [2, 4] ])                                # in both these cases output is same because MatMul converts (b, M) into Matrix object hence when either we take complex_transpose or matrix_transpose both give the same output.
```
similarly if i take `MatMul( mat_trans(b), M )` then `matrix_transpose` takes `b` as a scalar and hence returns `b` itself hence finally `b * M` is returned . But for `complex_trans(b)` it is returned as `complex_transpose(b)` but when `MatMul` is applied `complex_trans(b)`  it is taken as a scalar and hence `complex_trans(b) * M` is returned.

I find this way of computing very valid. Even conditions are satisfied for the `MatrixSymbol` the way they should be. Since `MatrixSymbol` is not a scalar it is a Matrix object.
What wrong do you find in this ? Please can you look.

**Comment by jksuom** [comment]
I think that we are trying to do something that is impossible to do before some kind of structure is added to non-commutative symbols.

``MatMul`` was originally developed as a variant of ``Mul`` that would check that the shapes of the factor matrices were compatible. It was later extended to allow scalar factors that would be separated from the actual matrices before checking the shapes. This is possible because commutative scalars can always be moved to any position in a matrix product.

With non-commutative factors the situation is different. They cannot be treated independently of the proper matrix factors. And the problem is that they have no shape. One possible solution could be that they were thought of as matrices of unknown shape. Then the ``matmul.validate`` routine should be modified to accept an unknown shape assuming, of course, that such an attribute was added to non-commutative symbols. This means that they would become very much like ``MatrixSymbol``s with indefinite shape.

In summary: Instead of trying to use arbitrary non-commutative symbols in matrix products one should only accept a special class that has a shape attribute with possibly indefinite values. Also ``validate`` should be extended to accept such values.
